src/Features/Core/Portable/Completion/Providers/AbstractPreprocessorCompletionProvider.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    internal abstract class AbstractPreprocessorCompletionProvider : LSPCompletionProvider
+    {
+        protected abstract Task<SyntaxContext> CreateContextAsync(Workspace workspace, SemanticModel semanticModel, int position, System.Threading.CancellationToken cancellationToken);
+
+        public sealed override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var originatingDocument = context.Document;
+            var position = context.Position;
+
+            var semanticModel = await originatingDocument.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+
+            var solution = originatingDocument.Project.Solution;
+            var syntaxContext = await CreateContextAsync(solution.Workspace, semanticModel, position, cancellationToken).ConfigureAwait(false);
+            if (!syntaxContext.IsPreProcessorExpressionContext)
+                return;
+
+            using var _ = PooledHashSet<string>.GetInstance(out var preprocessorNames);
```

Comment:
Preprocessor directives are case sensitive in C# but insensitive in VB. Need to define a hook here to let the concrete type specify the case sensitivity. Otherwise in VB we could end up with incorrect duplicates.

Reply:
sure.

Reply:
done.